crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -1207,14 +1209,60 @@ fn handle_dict_unpacking_comment<'a>(
     .skip_while(|token| token.kind == SimpleTokenKind::RParen);
 
     // if the remaining tokens from the previous node are exactly `**`,
-    // re-assign the comment to the one that follows the stars
+    // re-assign the comment to the one that follows the stars.
     if tokens.any(|token| token.kind == SimpleTokenKind::DoubleStar) {
         CommentPlacement::leading(following, comment)
     } else {
         CommentPlacement::Default(comment)
     }
 }
 
+/// Handles comments around the `:` in a key-value pair:
+///
+/// ```python
+/// {
+///     key  # dangling
+///     :  # dangling
+///     # dangling
+///     value
+/// }
+/// ```
+fn handle_key_value_comment<'a>(
+    comment: DecoratedComment<'a>,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    debug_assert!(matches!(
+        comment.enclosing_node(),
+        AnyNodeRef::ExprDict(_) | AnyNodeRef::ExprDictComp(_)
+    ));
+
+    let (Some(following), Some(preceding)) = (comment.following_node(), comment.preceding_node())
+    else {
+        return CommentPlacement::Default(comment);
+    };
+
+    // Ensure that the comment is between the key and the value by detecting the colon:
```

Comment:
I wonder if this gets comments wrong then the value or `key` are perenthesized. 

I would be leaning towards making comments before `:` trailing comments of `key` and all other comments leading comments of value + forcing a hard line break in the dictionary formatting when the value has any leading comments.

Reply:
It does in some cases but not all -- some cases are already handled by `handle_parenthesized_comment`. I will fix the others.

Reply:
I guess I can try this but candidly it feels more complicated. I will almost certainly need to change the `trailing_comments` implementation -- otherwise, this:
```python
{
    x:  # comment
    y
}
```

Will get formatted as:

```python
{
    x:# comment
    y
}
```

Let's see how it goes.

Reply:
We will also have to be able to differentiate between these cases when formatting the dictionary:

```python
{
    x:  # comment
    y
}
{
    x: (  # comment
        y
    )
}
```

Since in the first case, we don't want to insert any spaces after the `:` (`trailing_comments` would need to do this), whereas in the second case we do. I like that the current solution gets this right automatically, since the parenthesized comment is a leading comment but the unparenthesized comment is dangling.

Reply:
I'm actually not quite sure how to support that second case.

Reply:
Unless we are comfortable making `# comment` an own-line comment? Pushing it to the next line? But that feels off.